url89849c8c5b_timmolter_XChart_tgz-pJ8-org_knowm_xchart_standalone_Example1J8/src/org/knowm/xchart/CSVExporter.java

```diff
@@ -0,0 +1,200 @@
+  }
+
+  /**
+   * Export a XYChart series into rows in a CSV file.
+   *
+   * @param series
+   * @param path2Dir - ex. "./path/to/directory/" *make sure you have the '/' on the end
+   */
+  public static void writeCSVRows(XYSeries series, String path2Dir) {
+
+    File newFile = new File(path2Dir + series.getName() + ".csv");
+    Writer out = null;
+    try {
+
+      out = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(newFile), "UTF8"));
+      String csv = join(series.getXData(), ",") + System.getProperty("line.separator");
+      out.write(csv);
+      csv = join(series.getYData(), ",") + System.getProperty("line.separator");
+      out.write(csv);
+      if (series.getExtraValues() != null) {
+        csv = join(series.getExtraValues(), ",") + System.getProperty("line.separator");
+        out.write(csv);
+      }
+
+    } catch (Exception e) {
+      e.printStackTrace();
+    } finally {
+      if (out != null) {
+        try {
+          out.flush();
+          out.close();
+        } catch (IOException e) {
+          // NOP
+        }
+      }
+    }
+  }
+
+  /**
+   * Joins a collection into an entire row of comma separated values.
+   *
+   * @param collection
+   * @param separator
+   * @return
+   */
+  private static String join(Collection<? extends Object> collection, String separator) {
+
+    if (collection == null) {
+      return null;
+    }
+    Iterator<? extends Object> iterator = collection.iterator();
+    // handle null, zero and one elements before building a buffer
+    if (iterator == null) {
+      return null;
+    }
+    if (!iterator.hasNext()) {
+      return "";
+    }
+    Object first = iterator.next();
+    if (!iterator.hasNext()) {
+      return first == null ? "" : first.toString();
+    }
+
+    // two or more elements
+    StringBuffer buf = new StringBuffer(256); // Java default is 16, probably too small
+    if (first != null) {
+      buf.append(first);
+    }
+
+    while (iterator.hasNext()) {
+      if (separator != null) {
+        buf.append(separator);
+      }
+      Object obj = iterator.next();
+      if (obj != null) {
+        buf.append(obj);
+      }
+    }
+    return buf.toString();
+
+  }
+
+  /**
+   * Export all XYChart series as columns in separate CSV files.
+   *
+   * @param chart
+   * @param path2Dir
+   */
+  public static void writeCSVColumns(XYChart chart, String path2Dir) {
+
+    for (XYSeries xySeries : chart.getSeriesMap().values()) {
+      writeCSVColumns(xySeries, path2Dir);
+    }
+  }
+
+  /**
+   * Export a Chart series in columns in a CSV file.
+   *
+   * @param series
+   * @param path2Dir - ex. "./path/to/directory/" *make sure you have the '/' on the end
+   */
+  public static void writeCSVColumns(XYSeries series, String path2Dir) {
+
+    File newFile = new File(path2Dir + series.getName() + ".csv");
+    Writer out = null;
+    try {
+
+      out = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(newFile), "UTF8"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 151, 154, 182, 187, 188, 190. The resource is closed at line 191. There are other execution paths that do not contain closure statements, for example, when an exception is thrown by Writer.write, Writer.flush. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).